chapel-lang/chapel
# compiler/next: Get all but a few primers to compile with new parser

This is for `compiler/next`.

Get all but a few primers (`chpldoc`, `FFTWlib`, `LAPACKlib`) to
compile with the new frontend via the `--compiler-library-parser`
flag.

Add a few new UAST nodes: `Scan`, `TypeQuery`, and `VarArgFormal`.

Adjust the `TupleDecl` node so that it can also contain `Formal`.

Adjust the `Function::formals()` iterator and accessor to yield
`Decl` instead of `Formal`.

Adjust the conversion routines in `compiler/passes/convert-uast.cpp`
as needed to translate new UAST nodes as well as new combinations
of existing nodes (such as edge cases for array types).

Wire up a bunch of productions in the new parser grammar such as
for scans, varargs, loop expressions with filters, and more.

Remove redundant logic for setting config vars in the production
compiler. Reuse this function in the conversion pass for now.

Add a bunch of TODOs for functionality in the conversion pass
powered by the production compiler that should be rewritten
in the new compiler when possible.

Reviewed by @mppf, @arezaii. Thanks!

TESTING:

- [x] All tests with `linux64`, `COMM=none` to verify config change is OK
- [x] All primers except `FFTWlib`, `LAPACKlib`, `chpldoc` with
        `--compiler-library-parser` flag

FUTURE WORK:

- Handle pragmas and conversion of pragmas (ideally using an approach
  like that in #18619)
- Tackle all remaining TODOEXPR in the the new grammar file
- Pick another larger group of tests to run next
- Add verbosity levels to errors (e.g. `noteWarning`)
- Begin iterating through all production compiler `build` routines
  (used by the old parser) and ensuring that their warnings and
  errors also occur in the new parser
- Consider traversing the tree after parsing (say while assigning
  postorder IDs) and emitting some parse-time errors in order to
  decouple these errors from the parser proper
- Begin to selectively parse the internal modules with the new parser
- Add machinery to parse and look up config var values

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>
